### PR TITLE
Math Distillation Task

### PR DESCRIPTION
**Pull request by sanchez-alex**
Example runs:
1. [Multi-Arith CoT](https://ml.azure.com/experiments/id/a53977ba-c89d-47e1-8471-135325242dd1/runs/good_grass_kv1d246yvd?wsid=/subscriptions/75703df0-38f9-4e2e-8328-45f6fc810286/resourcegroups/rg-sasumai/workspaces/sasum-westus3-ws&tid=7f292395-a08f-4cc0-b3d0-a400b023b0d2&pipelineChildMetrics=true)
2. [Multi-Arith Zero Shot](https://ml.azure.com/experiments/id/c5eaf91e-f3b3-4313-b0cb-f78efb5a500b/runs/stoic_knot_k10y8tsvzj?wsid=/subscriptions/75703df0-38f9-4e2e-8328-45f6fc810286/resourcegroups/rg-sasumai/workspaces/sasum-westus3-ws&tid=7f292395-a08f-4cc0-b3d0-a400b023b0d2&pipelineChildMetrics=true)

### CODE REVIEW

**Review by hakotesova — DISMISSED** [review]

File: assets/training/distillation/src/common/constants.py [hunk 1]
```diff
@@ -53,7 +53,7 @@
 DEFAULT_SUCCESS_RATIO = 0.7
 DEFAULT_REQUEST_BATCH_SIZE = 10
 MAX_BATCH_SIZE = 100
-MIN_RECORDS_FOR_FT = 65
+MIN_RECORDS_FOR_FT = 40
```
Comment (skasturi) [inline assets/training/distillation/src/common/constants.py:56]: Could we please verify this for other tasks that are already released?

**Comment by babu-namburi** [comment]
@sanchez-alex 
Can you add validation for the MATH task in the function `_validate_record_by_task`

File: assets/training/distillation/src/common/constants.py (hunk 1 above)
Comment (sanchez-alex) [reply assets/training/distillation/src/common/constants.py]: Reverted back to original and allowed the min records to be lower for math tasks as that has been tested already

**Review by hakotesova — APPROVED** [review]

File: assets/training/distillation/src/generate_data.py [hunk 2]
```diff
@@ -366,14 +365,18 @@ def process_request(idx: str, data: dict, url: str, endpoint_key: str):
             messages = normalize_messages(messages)
             last_status_code = None
             synthetic_responses = []
+            inference_data = []
             for turn_id, message in enumerate(messages):
                 role = message["role"]
                 if role == "system":
-                    synthetic_responses.append(process_system_prompt(message))
+                    # Data for fine-tune job should not include CoT prompt
```
Comment (babu-namburi) [inline assets/training/distillation/src/generate_data.py:372]: @visahan-24 
Can you please check this change.

File: assets/training/distillation/src/generate_data.py [hunk 3]
```diff
@@ -389,11 +392,15 @@ def process_request(idx: str, data: dict, url: str, endpoint_key: str):
                         break
                     response_data = response.json()
                     # response content should be structured as below for a successful vllm response
-                    prediction_result = response_data["choices"][0]["message"][
-                        "content"
-                    ].strip()
+                    prediction_result = response_data["choices"][0]["message"]["content"].strip()
+
+                    # For CoT prompts, need to remove the reasoning and only use the answer
+                    if enable_cot and data_generation_task_type != DataGenerationTaskType.CONVERSATION:
```
Comment (babu-namburi) [inline assets/training/distillation/src/generate_data.py:398]: nit: Can we move this to a separate function which post process the response from the model. So that we can include post processing for other tasks without disturbing this function.

File: assets/training/distillation/src/generate_data.py (hunk 2 above)
Comment (visahan-24) [reply assets/training/distillation/src/generate_data.py:372]: @babu-namburi - This change looks good

File: assets/training/distillation/src/generate_data.py [hunk 4]
```diff
@@ -389,11 +392,15 @@ def process_request(idx: str, data: dict, url: str, endpoint_key: str):
...
+                        key = SystemPrompt.get_response_key(data_generation_task_type)
+                        prediction_result = json.loads(prediction_result)[key]
+
                     synthetic_responses.append(
-                        {"role": "assistant", "content": prediction_result}
+                        {"role": "assistant", "content": str(prediction_result)}
```
Comment (novaturient95) [inline assets/training/distillation/src/generate_data.py:403]: nit: isn't prediction_result always str?

File: assets/training/distillation/src/validate_pipeline.py [hunk 5]
```diff
@@ -186,7 +186,7 @@ def _validate_inference_parameters(self):
 
     def _validate_number_of_records(self, size: int):
         """Validate number of records in the dataset."""
-        if size < MIN_RECORDS_FOR_FT:
+        if size < MIN_RECORDS_FOR_FT and self._args.data_generation_task_type != DataGenerationTaskType.MATH:
```
Comment (novaturient95) [inline assets/training/distillation/src/validate_pipeline.py:189]: Could you please share reasoning behind this check? AFAIU, even math would need to follow min row req.

Comment (sanchez-alex) [reply assets/training/distillation/src/validate_pipeline.py:189]: I used 3 math datasets to test this. Some math datasets don't have a validation set so it is taken from the training data. The data science team had a 90-10 split for when no validation set was available. For one dataset, the Multi-Arith dataset, the number of entries for validation following the 90-10 split was 42 records. I was able to finetune the llama-3.1-8b model successfully and there were improvements in the model's accuracy as well even with this small amount of validation data.

File: assets/training/distillation/src/generate_data.py (hunk 4 above)
Comment (sanchez-alex) [reply assets/training/distillation/src/generate_data.py:403]: For math datasets, a number is asked to be returned, and a number value is a valid value in a JSON, so the str is to keep everything consistent

File: assets/training/distillation/src/generate_data.py (hunk 3 above)
Comment (sanchez-alex) [reply assets/training/distillation/src/generate_data.py:398]: Can add this change in a separate PR

**Review by hakotesova — APPROVED** [review]

File: assets/training/distillation/src/common/constants.py [hunk 6]
```diff
@@ -69,16 +70,7 @@
...
-COT_SYSTEM_PROMPT = (
-    "You are a helpful assistant. "
-    "Write out in a step by step manner your reasoning about the answer using no more than 80 words. "
-    "Based on the reasoning, produce the final answer. "
-    "Your response should be in JSON format without using any backticks. "
-    "The JSON is a dictionary whose keys are 'reason' and 'answer_choice'. "
-    "Always generate a syntactically correct JSON without using markdown and any additional words. "
-)
-
+# CHAIN OF THOUGHT (COD)
```
Comment (sharvin2187) [inline assets/training/distillation/src/common/constants.py:73]: nit: COD -> COT

**Review by msp8955 — APPROVED** [review]

File: assets/training/distillation/components/pipeline/spec.yaml [hunk 7]
```diff
@@ -257,7 +259,7 @@ jobs:
 
   oss_distillation_generate_data:
     type: command
-    component: azureml:oss_distillation_generate_data:0.0.5
+    component: azureml:oss_distillation_generate_data:0.0.6
```
Comment (sharvin2187) [inline assets/training/distillation/components/pipeline/spec.yaml:262]: we recently faced some issues with versioning , the correct steps to follow would be:
1. just before release, check the version in azureml registry, and make the increment for all components/pipelines which will be updated due to the release.
2. build all the affected components with the updated versions, release them together.

not immediately relevant to the PR, but will become important during release.

Comment (sanchez-alex) [reply assets/training/distillation/components/pipeline/spec.yaml:262]: Sounds good, will verify all versions in prod after this PR

File: assets/training/distillation/src/common/constants.py (hunk 6 above)
Comment (sanchez-alex) [reply assets/training/distillation/src/common/constants.py]: Removed comment. Author of COD_SYSTEM_PROMPT did not have add a comment and not sure what D stands for in COD

**Review by msp8955 — APPROVED** [review]

**Review by hakotesova — APPROVED** [review]
